### PR TITLE
[CL]: add underlying lock ID func + helpers (2/5)

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

This PR is one of five PRs in an effort to make #4854 more reviewable.

The scope of this PR is limited to adding the underlying lock ID functions as well as any helpers. 

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/concentrated-liquidity/lp.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp.go:436]: isLockMature test gets added in the next PR here https://github.com/osmosis-labs/osmosis/pull/4883/files#diff-8a43021241fc488af68dd6cdd7c762a85476f3aa8e44248784e9265c9ef0bbdf

This is because the isLockMature test relies on createFullRangePosition unlocking method that is not introduced until the next PR

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:466]: Is this new index added to genesis? If not, can you please do so in this PR

File: x/concentrated-liquidity/genesis.go
Comment (p0mvn) [inline x/concentrated-liquidity/genesis.go:68]: Could you elaborate? I'm not following why this is zero. I think we should have lock id to position mapping id be exported

File: x/concentrated-liquidity/lp.go
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:254]: discussed offline to remove `underlyingLockId` from `updatePosition` since it's always set to 0

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:187]: Please cover this by tests. I think this one is important since there are a few indexes get updated here

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [reply x/concentrated-liquidity/position.go:466]: Discussed adding in separate PR but please add it to the issue and a TODO with link to issue in code

File: x/concentrated-liquidity/types/keys.go
Comment (p0mvn) [inline x/concentrated-liquidity/types/keys.go:218]: Can we do a strict length check? So that this panics if the number of parts is not as expected

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:222]: I think we should be covering this logic with tests as well. Having such areas untested would make any potential refactors much easier because a small accidental change would be caught by the direct unit test

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:415]: nit: I think it would be useful to format position id here as well

Comment (p0mvn) [inline x/concentrated-liquidity/position.go:415]: Not seeing test cases added for this logic, please add them

File: x/concentrated-liquidity/lp.go
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:161]: Would like to see test cases added for this as well

**Review by p0mvn — COMMENTED** [review]
The logic LGTM.

Would like to make sure that we have sufficient test coverage for all core logic before approving. Pointed out areas that I think should be covered by tests

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go:470]: The issue was already created, but added the TODO in code here https://github.com/osmosis-labs/osmosis/pull/4882/commits/8436cd092f70047d37b9133211a06a087320ac28

File: x/concentrated-liquidity/lp.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go]: https://github.com/osmosis-labs/osmosis/pull/4882/commits/e6fc42b231be2553e47e0676d23ec738d2f4852a

File: x/concentrated-liquidity/genesis.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/genesis.go:68]: I believe I addressed this in the following comment https://github.com/osmosis-labs/osmosis/pull/4882/#discussion_r1162274554

If this is related, please resolve this.

File: x/concentrated-liquidity/types/keys.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/types/keys.go]: https://github.com/osmosis-labs/osmosis/pull/4882/commits/5cb0a2d99233a3d9c1a32edc3666340b99759ad3

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go:191]: Added SetPosition test here https://github.com/osmosis-labs/osmosis/pull/4882/commits/77fad9302ff0238def4cec1b1dde1aa28800a994

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go]: The test case for validatePositionsAndGetTotalLiquidity is included in the fungify charge position test here https://github.com/osmosis-labs/osmosis/blob/97caf31bafde3d635cb7e4001d6222d441619c5d/x/concentrated-liquidity/position_test.go#L570

File: x/concentrated-liquidity/lp.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go:161]: I am going to implement this into the next PR right now and link it here (in order to properly implement this test I want to have CreatePositionUnlocking and Locked methods)

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/position.go:487]: Test added in next PR here https://github.com/osmosis-labs/osmosis/pull/4883/commits/d5715afd11c67b4e4c5cdeca67286d92a2f69daf

File: x/concentrated-liquidity/lp.go
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:277]: This isn't right, why is this always set to zero? What's the purpose of this parameter?

This is the only place that this method is called

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:67]: I'm guessing `noUnderlyingLockId` should be set here and `underlyingLockId` be removed from `initOrUpdatePosition` params?

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:445]: Can we move this `cl/pool` prefix somewhere more visible? E.g. `keys.go`?

File: x/concentrated-liquidity/types/keys.go
Comment (p0mvn) [inline x/concentrated-liquidity/types/keys.go:224]: It might also be useful to validate that denom matches the desired prefix

File: x/concentrated-liquidity/position_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/position_test.go:1022]: Nice

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:419]: Please update godoc for this keeper

File: x/concentrated-liquidity/position.go
Comment (p0mvn) [inline x/concentrated-liquidity/position.go:487]: WDYT about?

Slightly worried that we get false positive when some random error occurs

```suggestion
func (k Keeper) positionHasUnderlyingLockInState(ctx sdk.Context, positionId uint64) (bool, error) {
	// Get the lock ID for the position.
	_, err := k.GetPositionIdToLock(ctx, positionId)
	if err == nil || !errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
		return true, nil
	}
	return false, err
}
```

File: x/concentrated-liquidity/lp.go
Comment (p0mvn) [reply x/concentrated-liquidity/lp.go:161]: Not going to block this PR but in the future, please aim to add tests directly in the PR so that the reviewer can run them and make sure that edge cases are covered sufficiently

**Review by p0mvn — APPROVED** [review]
LGTM once comments are addressed. Nice work

File: x/concentrated-liquidity/lp.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go:161]: I could not implement this test without the Unlocking and Locked methods that were added in the next PR. If I added them here, the next PR would have had nothing to review and would increased the scope of this PR

File: x/concentrated-liquidity/lp.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/lp.go]: You are right, updated here https://github.com/osmosis-labs/osmosis/pull/4882/commits/5855df51844dfb23cf09f734f05ca6d328afb1a1

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go]: Yes, done here https://github.com/osmosis-labs/osmosis/pull/4882/commits/5855df51844dfb23cf09f734f05ca6d328afb1a1

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go]: yes, moved to cl.go in the types package here https://github.com/osmosis-labs/osmosis/pull/4882/commits/044b4b923a62996cff40bb56caed5d108fbc1597

File: x/concentrated-liquidity/types/keys.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/types/keys.go]: added prefix validation here https://github.com/osmosis-labs/osmosis/pull/4882/commits/cc1624e9e5898a5ad8026bdd9f31a71e4851aa19

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go]: godoc updated here https://github.com/osmosis-labs/osmosis/pull/4882/commits/aa7bbd23185ff4bc94500dbb6cc1cca7cc1c60b8

File: x/concentrated-liquidity/position.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/position.go]: added error here https://github.com/osmosis-labs/osmosis/pull/4882/commits/d0f00c83f04c2d6fd37a60de13567bcca7841c61

**Pull request merged**
